pkg/agent/cache/manager/manager.go

```diff
@@ -0,0 +1,302 @@
+package manager
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+
+	"crypto/tls"
+	"crypto/x509"
+	"github.com/sirupsen/logrus"
+	spiffe_tls "github.com/spiffe/go-spiffe/tls"
+	"github.com/spiffe/go-spiffe/uri"
+	"github.com/spiffe/spire/pkg/agent/cache"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/api/node"
+	proto "github.com/spiffe/spire/proto/common"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"io"
+	"sync"
+	"time"
+)
+
+type Manager interface {
+	UpdateCache(stop chan struct{})
+	fetchSVID(requests []EntryRequest, nodeClient node.NodeClient, wg *sync.WaitGroup)
+	getGRPCConn(svid []byte, key *ecdsa.PrivateKey) (*grpc.ClientConn, error)
+	expiredCacheEntryHandler(stop chan struct{})
+}
+
+type EntryRequest struct {
+	CSR   []byte
+	entry cache.CacheEntry
+}
+
+type manager struct {
+	managedCache     cache.Cache
+	serverCerts      []*x509.Certificate
+	serverAddr       string
+	serverSPIFFEID   string
+	errorCh          chan error
+	entryRequestCh   chan map[string][]EntryRequest
+	regEntriesCh     chan []*proto.RegistrationEntry
+	CacheEntryCh     chan cache.CacheEntry
+	spiffeIdEntryMap map[string]cache.CacheEntry
+	baseSVID         []byte
+	baseSVIDKey      *ecdsa.PrivateKey
+	baseSPIFFEID     string
+	regEntries       []*proto.RegistrationEntry
+	log              logrus.FieldLogger
+}
+
+func NewManager(managedCache cache.Cache,
+	serverCerts []*x509.Certificate,
+	serverSPIFFEID string,
+	serverAddr string,
+	errorCh chan error,
+	baseSVID []byte, baseSVIDKey *ecdsa.PrivateKey, regEntries []*proto.RegistrationEntry, logger logrus.FieldLogger) manager {
+	cert, err := x509.ParseCertificate(baseSVID)
+	if err != nil {
+		errorCh <- err
+	}
+	basespiffeID, err := uri.GetURINamesFromCertificate(cert)
+
+	return manager{
+		managedCache:   managedCache,
+		serverCerts:    serverCerts,
+		serverSPIFFEID: serverSPIFFEID,
+		serverAddr:     serverAddr,
+		baseSVID:       baseSVID,
+		baseSVIDKey:    baseSVIDKey,
+		baseSPIFFEID:   basespiffeID[0],
+		regEntries:     regEntries,
+		errorCh:        errorCh,
+		log:            logger.WithField("subsystem_name", "cacheManager"),
+	}
+}
+
+func (m *manager) UpdateCache(stop chan struct{}) {
+	m.spiffeIdEntryMap = make(map[string]cache.CacheEntry)
+	m.entryRequestCh = make(chan map[string][]EntryRequest)
+	m.regEntriesCh = make(chan []*proto.RegistrationEntry)
+	m.errorCh = make(chan error)
+	m.CacheEntryCh = make(chan cache.CacheEntry)
```

Comment:
I don't think this stuff should be modifying `m`.  Anything stopping `UpdateCache()` from being called twice?

Why is `CacheEntryCh` exported?